File: services/cookbook/core/src/main/java/family/haschka/wolkenschloss/cookbook/job/ImportJob.java
package family.haschka.wolkenschloss.cookbook.job;

import java.net.URI;
import java.util.UUID;

public record ImportJob(UUID jobId, URI order, State state, URI location, String error) {

    @Override
    public String toString() {
        return "ImportJob{" +
                "url='" + order + '\'' +
                ", jobId=" + jobId +
                ", state=" + state +
                ", location=" + location +
                ", error=" + error +
                '}';
    }

    public ImportJob complete(URI location, String error) {
        return new ImportJob(jobId, order, State.COMPLETED, location, error);
    }

    public ImportJob located(URI location) {
        return new ImportJob(jobId, order, State.INCOMPLETE, location, error);
    }

    public static ImportJob create(UUID jobId, URI order) {
        return new ImportJob(jobId, order, State.CREATED, null, null);
    }

    public ImportJob failed(Throwable failure) {
        return new ImportJob(jobId, order, State.FAILED, location, failure.getMessage());
    }
}

File: services/cookbook/core/src/main/kotlin/job/ImportJob.kt
package family.haschka.wolkenschloss.cookbook.job

import java.net.URI
import java.util.*

data class ImportJob(val jobId: UUID, val order: URI, val state: State, val location: URI, val error: String) {

    fun complete(location: URI): ImportJob =
        this.copy(location = location, state = State.COMPLETED)

    fun located(location: URI): ImportJob =
        this.copy(location = location, state = State.INCOMPLETE)

    fun failed(failure: Throwable) =
        this.copy(state = State.FAILED, error = failure.message ?: "Unknown failure")

    companion object {
        @JvmStatic
        fun create(jobId: UUID, order: URI): ImportJob =
            ImportJob(jobId, order, State.CREATED, NoLocation, "")

        @JvmStatic val NoLocation: URI = URI.create("#")
    }
}
